cmd/influxd/inspect/report_tsm/report_tsm.go

```diff
@@ -0,0 +1,356 @@
+overall cardinality in the file set by using the HLL++ algorithm. Exact 
+cardinalities can be determined by using the --exact flag.
+For each file, the following is output:
+	* The full filename;
+	* The series cardinality within the file;
+	* The number of series first encountered within the file;
+	* The min and max timestamp associated with TSM data in the file; and
+	* The time taken to load the TSM index and apply any tombstones.
+The summary section then outputs the total time range and series cardinality for 
+the fileset. Depending on the --detailed flag, series cardinality is segmented 
+in the following ways:
+	* Series cardinality for each organization;
+	* Series cardinality for each bucket;
+	* Series cardinality for each measurement;
+	* Number of field keys for each measurement; and
+	* Number of tag values for each tag key.`,
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			// Verify if shard dir
+			err := arguments.isShardDir(arguments.dir)
+			if arguments.detailed && err != nil {
+				return fmt.Errorf("--detailed only supported for shard dirs")
+			}
+
+			return arguments.Run(cmd)
+		},
+	}
+
+	cmd.Flags().StringVarP(&arguments.pattern, "pattern", "", "", "only process TSM files containing pattern")
+	cmd.Flags().BoolVarP(&arguments.exact, "exact", "", false, "calculate and exact cardinality count. Warning, may use significant memory...")
+	cmd.Flags().BoolVarP(&arguments.detailed, "detailed", "", false, "emit series cardinality segmented by measurements, tag keys and fields. Warning, may take a while.")
+
+	dir, err := fs.InfluxDir()
+	if err != nil {
+		panic(err)
+	}
+	dir = filepath.Join(dir, "engine/data")
+	cmd.Flags().StringVarP(&arguments.dir, "data-dir", "", dir, fmt.Sprintf("use provided data directory (defaults to %s).", dir))
+
+	return cmd
+}
+
+func (a *args) isShardDir(dir string) error {
+	name := filepath.Base(dir)
+	if id, err := strconv.Atoi(name); err != nil || id < 1 {
+		return fmt.Errorf("not a valid shard dir: %v", dir)
+	}
+
+	return nil
+}
+
+func (a *args) Run(cmd *cobra.Command) error {
+	// Create the cardinality counter
+	newCounterFn := newHLLCounter
+	estTitle := " (est)"
+	if a.exact {
+		estTitle = ""
+		newCounterFn = newExactCounter
+	}
+
+	totalSeries := newCounterFn()
+	tagCardinalities := map[string]counter{}
+	measCardinalities := map[string]counter{}
+	fieldCardinalities := map[string]counter{}
+
+	dbCardinalities := map[string]counter{}
+
+	start := time.Now()
+
+	tw := tabwriter.NewWriter(cmd.OutOrStdout(), 8, 2, 1, ' ', 0)
+	_, _ = fmt.Fprintln(tw, strings.Join([]string{"DB", "RP", "Shard", "File", "Series", "New" + estTitle, "Min Time", "Max Time", "Load Time"}, "\t"))
+
+	minTime, maxTime := int64(math.MaxInt64), int64(math.MinInt64)
+	var fileCount int
+	if err := a.walkShardDirs(a.dir, func(db, rp, id, path string) error {
+		if a.pattern != "" && strings.Contains(path, a.pattern) {
```

Comment:
Should this be:
```suggestion
		if a.pattern != "" && !strings.Contains(path, a.pattern) {
```
?